core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -48,20 +123,85 @@ import Pull._
   *
   * `raiseError` is caught by `handleErrorWith`:
   *   - `handleErrorWith(raiseError(e))(f) == f(e)`
+  *
+  * @tparam F[_] the type of functional effects that can be performed by this pull.
+  * @tparam O The type of outputs emitted by this Pull.
+  * @tparam R The type of carry-over result that is returned by this Pull, if it returns successfully.
   */
 sealed abstract class Pull[+F[_], +O, +R] {
 
-  /** Alias for `_.map(_ => o2)`. */
-  def as[R2](r2: R2): Pull[F, O, R2] = map(_ => r2)
+  /** Applies the result of this pull to `f` and returns the result.
+    *
+    * This method returns a new composed pull. The semantics of this pull
+    *
+    * - If `this` pull completes successfully with a result of type R, 
+    *   the `f` function is applied to the result to build a new Pull, 
+    *   and compilation continues by running that pull. The end return of 
+    *   the composed pull is that of the generated pull.
+    * - If this pull fails or is interrupted, the failure of interruption is 
+    *   the returned of the composed pull.
+    *
+    * The composed pull will first emit all the outputs emitted by `this` pull,
+    * and if successful will start emitting the outputs from the generated pull.
+    *
+    * This operation does not modify resource scope boundaries. The generated `post` 
+    * pull starts compilation on the same scope on which the compilation of this 
+    * pull (the prefix) ends. The composed pull will end on whatever scope the 
+    * compilation of the `post` pull does.
+    */
+  def flatMap[F2[x] >: F[x], O2 >: O, R2](kont: R => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
+    new Bind[F2, O2, R, R2](this) {
+      def cont(e: Terminal[R]): Pull[F2, O2, R2] =
+        e match {
+          case Succeeded(r) =>
+            try kont(r)
+            catch { case NonFatal(e) => Fail(e) }
+          case res @ Interrupted(_, _) => res
+          case res @ Fail(_)           => res
+        }
+    }
 
-  /** Returns a pull with the result wrapped in `Right`, or an error wrapped in `Left` if the pull has failed. */
-  def attempt: Pull[F, O, Either[Throwable, R]] =
-    map(r => Right(r)).handleErrorWith(t => Succeeded(Left(t)))
+  /** Lazily appends the given `post` pull, to be run after `this` pull.
```

Comment:
I think this documentation should be really lightweight with a simple reference to the `flatMap` definition. Describe this as an alias for flatMap and describe how it differs from `*>` re: laziness and allowing tail recursive pulls.

Reply:
OK, I don't feel strongly. Add a mention that this is the same as `flatMap(_ => that)` though.